torch/nn/parallel/distributed.py

```diff
@@ -388,6 +386,10 @@ def __init__(
             "doesn't have any parameter that requires a gradient."
         )
 
+        assert (
+            device_ids is None or len(device_ids) <= 1
```

Comment:
I think we are mixing up two different modes here.

1. There is a multi-device model mode where the parameters of the module span across multiple GPUs (basically some form of model parallelism). We want DDP to support this mode and in this mode, device_ids will be None.
2. There is a single process multi device mode, where the model itself is on a single device, but len(device_ids) > 1 indicating that a single DDP process should replicate this model across multiple GPUs. We would like to deprecate this mode.

Reply:
Thanks for the explanation! I totally ignored multi-device model mode here.

I think the assert condition is correct. Updated the error message here, to describe the above two modes separately.